apps/bot/src/services/streamstats/index.ts

```diff
@@ -0,0 +1,180 @@
+import { prisma, prismaQueue, storage } from "services/db";
+import { TwitchApi } from "services/twitch-api";
+import type { TTwitchApiChatter, TTwitchIrcContext } from "services/types";
+import { logger } from "utils/logger";
+
+export class StreamStatsGatherer {
+  public static instances = new Map<string, StreamStatsGatherer>();
+
+  private api: TwitchApi;
+
+  private activeStream: string | undefined = undefined;
+  private activeStreamId = -1;
+
+  constructor(private channel: string) {
+    this.api = TwitchApi.getInstance(channel);
+  }
+
+  public static getInstance(ch: string): StreamStatsGatherer {
+    const instance = StreamStatsGatherer.instances.get(ch);
+
+    if (instance) {
+      return instance;
+    }
+
+    const newInstance = new StreamStatsGatherer(ch);
+
+    StreamStatsGatherer.instances.set(ch, newInstance);
+    return newInstance;
+  }
+
+  private statsInterval: Timer | undefined = undefined;
+  private standbyInterval: Timer | undefined = undefined;
+
+  private async startGathering(startedAt: string, streamerId: string, interval = 30_000) {
+    logger.info(`${this.channel} is now live, starting gathering stats`);
+    storage.set(`${this.channel}_is_live`, true);
+
+    storage.set(this.messagesKey, 0);
+    storage.set(this.newChattersKey, 0);
+    this.activeStream = `${startedAt}@${this.channel}`;
+
+    const res = await prisma.streams.upsert({
+      where: { unique_id: this.activeStream },
+      create: {
+        started_at: startedAt,
+        channel: this.channel,
+        unique_id: this.activeStream,
+        streamer_id: streamerId,
+      },
+      update: { started_at: startedAt },
+    });
+
+    this.activeStreamId = res.id;
+
+    this.statsInterval = setInterval(() => {
+      this.handleGathering();
+    }, interval);
+  }
+
+  public async startStandby() {
+    logger.info(`${this.channel} is now offline, standby mode`);
+    storage.set(`${this.channel}_is_live`, false);
+    clearInterval(this.statsInterval);
+
+    this.standbyInterval = setInterval(async () => {
+      await this.handleStandby();
+    }, 60_000);
+
+    if (!this.activeStream) {
+      return;
+    }
+
+    const uniqueChatters = storage.get<string[]>(this.uniqueChattersKey);
+    storage.delete(this.uniqueChattersKey);
+
+    await prisma.streams.update({
+      where: { unique_id: this.activeStream },
+      data: {
+        finished_at: String(Date.now()),
+        uniqueChatters: uniqueChatters?.value.length,
+      },
+    });
+  }
+
+  private get messagesKey() {
+    return `${this.channel}_stats_messages_in_that`;
```

Comment:
I do have some memory of similar type of `get`'s in other classes, but I'm not 100% sure about it.  
But maybe, extract these string manipulations (if they repeat), and make a class that only handles these string manipulations and use this here, that would probably reduce the friction and one less responsibility on this class.

Ps.: I'll not flood this MR with the other occurrences of this type of thing 😄

Reply:
you can consider this one as local storage key, I keep it this way because of a multiton, otherwise it could be just constant but I need to narrow a key based on channel